src/main/scala/strawman/collection/Map.scala

```diff
@@ -67,20 +68,23 @@ trait MapOps[K, +V, +CC[X, Y] <: MapOps[X, Y, CC, _], +C <: MapOps[K, V, CC, C]]
   /** Collects all keys of this map in a set.
     * @return  a set containing all keys of this map.
     */
-  def keySet: Set[K] = new DefaultKeySet
+  def keySet: Set[K] = new KeySet
 
   /** The implementation class of the set returned by `keySet`.
     */
-  protected class DefaultKeySet extends Set[K] with Serializable {
-    def contains(key: K) = MapOps.this.coll.contains(key)
-    def iterator() = keysIterator()
-    override def size = toIterable.size
-    override def foreach[U](f: K => U) = keysIterator() foreach f
+  protected class KeySet extends Set[K] with GenKeySet {
     def iterableFactory: IterableFactory[Set] = Set
-    def empty: Set[K] = iterableFactory.empty
-    def diff(that: Set[K]): Set[K] = fromSpecificIterable(iterableFactory.fromIterable(toIterable).diff(that))
-    protected[this] def fromSpecificIterable(coll: Iterable[K]): Set[K] = iterableFactory.fromIterable(coll)
+    protected[this] def fromSpecificIterable(coll: Iterable[K]): Set[K] = fromIterable(coll)
     protected[this] def newSpecificBuilder(): Builder[K, Set[K]] = iterableFactory.newBuilder()
+    def diff(that: Set[K]): Set[K] = fromSpecificIterable(this).diff(that)
```

Comment:
Basically we implement the `diff` operation by delegating to the `IterableFactory` defined above.

`fromSpecificIterable(this)` builds a `Set[K]` by using the `IterableFactory` defined a few lines above. This creates a copy of this keyset. Then we call its `diff` method.